src/nfa/nfa.cc

```diff
@@ -569,34 +559,44 @@ void Nfa::trim_inplace(StateToStateMap* state_map)
     }
 }
 
-void Nfa::trim_reverting(StateToStateMap* state_map)
+void Nfa::trim_inplace(StateToStateMap* state_map)
 {
 #ifdef _STATIC_STRUCTURES_
-    static NumberPredicate<State> useful_states(false);
+    std::vector<char> useful_states{ get_useful_states() };
     useful_states.clear();
     useful_states = get_useful_states();
 #else
-    NumberPredicate<State> useful_states{ get_useful_states() };
+    std::vector<char> useful_states{ get_useful_states() };
 #endif
 
-    initial.defragment(useful_states);
-    final.defragment(useful_states);
-    //this renaming can be used instead of the state map, it is computed there anyway, so far not used.
-    std::vector<State> renaming = delta.defragment(useful_states);
+    std::vector<State> renaming(useful_states.size());
 
-    // TODO : this is actually ony used in one test, remove state map?
+    State j=0;
+    for(State i = 0; i<useful_states.size(); i++) {
+        if (useful_states[i]) {
+            renaming[i] = j;
+            j++;
+        }
+    }
+
+    delta.defragment(useful_states, renaming);
+
+    auto is_state_useful = [&useful_states](State q){return useful_states[q];};
+    initial.filter(is_state_useful);
+    final.filter(is_state_useful);
+    auto rename_state = [&renaming](State q){return renaming[q];};
+    initial.rename(rename_state);
+    final.rename(rename_state);
+    initial.truncate();
+    final.truncate();
+
+    // TODO : this is actually only used in one test, remove state map?
     if (state_map) {
```

Comment:
That is the cost of maintaining the interface with features that we might not use right now.

We said that it could be used in some algorithms later. So far, we do not have a use for it, but maybe someone in the future will. What I am trying to say is that we need to decide whether we want this in the library public interface for someone to use (as we agreed on earlier), or not. I am fine with either decision.